### ISSUE TITLE
Fix CORS errors

### ISSUE THREAD

**Issue by philhanna** [philhanna/cwcomp#42]
Getting this error on the client when I try to fetch the results of a login request:
```
  Access to fetch at 'http://localhost:5053/login' from origin 'http://localhost:5173'
  has been blocked by CORS policy: Response to preflight request doesn't pass
  access control check: No 'Access-Control-Allow-Origin' header is present on
  the requested resource. If an opaque response serves your needs, set the request's mode
  to 'no-cors' to fetch the resource with CORS disabled.
```
This is from the so-called "preflight request", which is an HTTP OPTIONS call.  I need to find out what each side of the conversation is expecting.  The React.js code is this:
```js
fetch('http://localhost:5053/login', {
    method: 'POST',
    headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
    },
    body: JSON.stringify({ username: 'username', password: 'password' })
})
    .then(response => response.json())
    .then(data => console.log(data))
    .catch(error => console.error(error));

```

**Comment by philhanna** [comment]
Using a debug server:
```go

type loggingHandler struct {
	handler http.Handler
}

func (h loggingHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	// Log incoming headers
	log.Printf("Incoming method:  %q\n", r.Method)
	log.Printf("Incoming headers: %v", dumpHeaders(r.Header))

	// Call the wrapped handler
	h.handler.ServeHTTP(w, r)

	// Log outgoing headers
	log.Printf("Outgoing headers: %v", dumpHeaders(w.Header()))
}

func main() {
	// Create a new handler that logs headers and wraps the existing handler
	handler := loggingHandler{http.DefaultServeMux}

	// Start the server
	log.Fatal(http.ListenAndServe(":5053", handler))
}
```

**Comment by philhanna** [comment]
This is what I got using `"credentials": "omit"`, which is the default:
```
2023/06/22 19:38:42 debug_server.go:20: Incoming method:  "OPTIONS"
2023/06/22 19:38:42 debug_server.go:21: Incoming headers: 
    "Connection": "keep-alive"
    "Origin": "http://localhost:5173"
    "Sec-Fetch-Mode": "cors"
    "Sec-Fetch-Site": "same-site"
    "Referer": "http://localhost:5173/"
    "Accept-Encoding": "gzip, deflate, br"
    "Accept": "*/*"
    "Access-Control-Request-Method": "POST"
    "Access-Control-Request-Headers": "access-control-allow-origin,content-type,credentials"
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
    "Sec-Fetch-Dest": "empty"
    "Accept-Language": "en-GB,en-US;q=0.9,en;q=0.8,ru;q=0.7"


2023/06/22 19:38:42 debug_server.go:27: Outgoing headers: 
    "Content-Type": "text/plain; charset=utf-8"
    "X-Content-Type-Options": "nosniff"
```

Sure enough, the outgoing response does not contain the 'Access-Control-Allow-Origin' header the client is complaining about.

**Comment by philhanna** [comment]
Now I'll try it with `"credentials": "same-origin"`:
```
2023/06/22 19:54:51 debug_server.go:20: Incoming method:  "OPTIONS"
2023/06/22 19:54:51 debug_server.go:21: Incoming headers: 
    "Connection": "keep-alive"
    "Access-Control-Request-Method": "POST"
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
    "Sec-Fetch-Mode": "cors"
    "Accept-Language": "en-GB,en-US;q=0.9,en;q=0.8,ru;q=0.7"
    "Accept": "*/*"
    "Access-Control-Request-Headers": "access-control-allow-origin,content-type,credentials"
    "Origin": "http://localhost:5173"
    "Sec-Fetch-Site": "same-site"
    "Sec-Fetch-Dest": "empty"
    "Referer": "http://localhost:5173/"
    "Accept-Encoding": "gzip, deflate, br"


2023/06/22 19:54:51 debug_server.go:27: Outgoing headers: 
    "Content-Type": "text/plain; charset=utf-8"
    "X-Content-Type-Options": "nosniff"


```

**Comment by philhanna** [comment]
Finally, with `"credentials": "include"`:
```
2023/06/22 19:59:15 debug_server.go:20: Incoming method:  "OPTIONS"
2023/06/22 19:59:15 debug_server.go:21: Incoming headers: 
    "Accept-Encoding": "gzip, deflate, br"
    "Accept-Language": "en-GB,en-US;q=0.9,en;q=0.8,ru;q=0.7"
    "Connection": "keep-alive"
    "Access-Control-Request-Method": "POST"
    "Origin": "http://localhost:5173"
    "Sec-Fetch-Mode": "cors"
    "Sec-Fetch-Dest": "empty"
    "Referer": "http://localhost:5173/"
    "Accept": "*/*"
    "Access-Control-Request-Headers": "access-control-allow-origin,content-type,credentials"
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
    "Sec-Fetch-Site": "same-site"


2023/06/22 19:59:15 debug_server.go:27: Outgoing headers: 
    "Content-Type": "text/plain; charset=utf-8"
    "X-Content-Type-Options": "nosniff"
```

**Comment by philhanna** [comment]
According to perplexity, the server needs to respond to the preflight request with the following headers:

- **`Access-Control-Allow-Origin`**: This header indicates which origins are
  allowed to access the resource. It can be set to a specific origin or to * to
  allow any origin to access the resource.

- **`Access-Control-Allow-Methods`**: This header indicates which HTTP methods
  are allowed to access the resource. It can be set to a specific method or to
  a list of methods separated by commas.

- **`Access-Control-Allow-Headers`**: This header indicates which headers are
  allowed to be sent with the actual request. It can be set to a specific
  header or to a list of headers separated by commas.

- **`Access-Control-Max-Age`**: This header indicates how long the preflight
  response can be cached, in seconds. This can help reduce the number of
  preflight requests that are sent.

If the server does not respond with these headers, or if the headers are not
set correctly, the browser will not allow the actual request to be sent

**Comment by philhanna** [comment]
## Access-Control-Allow-Origin ##
In production, I'll probably need to [whitelist](https://en.wikipedia.org/wiki/Whitelist) some hosts and ports, but for now, I'll just use what came in on the `Origin` header, the URL of the client.

**Comment by philhanna** [comment]
The client gets an error message:
> Access to fetch at 'http://localhost:5053/login' from origin 'http://localhost:5173'
> has been blocked by CORS policy: Request header field credentials is not allowed
> by Access-Control-Allow-Headers in preflight response.

**Comment by philhanna** [comment]
Added that header.  That did it!

All I need to do now is incorporate that logic into the real handlers.
I'll open a new issue #43 to do that.
